### ISSUE TITLE
[Linux] maximizing and restoring the Notes window freezes / hangs the application

### ISSUE THREAD

**Issue by moritz1000** [nuttyartist/notes#611]
I am using Arch Linux & KDE.

https://github.com/nuttyartist/notes/assets/35174351/195d9a26-d931-48b6-a808-f7c98520acc3

Steps to reproduce:
1) Maximize Notes
2) Restore Notes to the original size

-> Freeze / Hang
-> Can't interact with the window anymore / type something

You need to kill the application and restart it in order to use it again.

___
It does not matter whether I am using the "use native window frame" option or not.

**Comment by zjeffer** [comment]
Hi, thanks for reporting. This is a known issue but we forgot to create an issue for this. 

The problem is somewhere in the newest Qt version (6.5), because the problem does not appear on Qt 6.4.3. We created a bug report here: https://bugreports.qt.io/browse/QTBUG-115459.

An alternative for you is to use the appimage instead, until this bug is fixed.

**Comment by nuttyartist** [comment]
@moritz1000 Can you please open a new issue for this problem?

**Comment by guihkx** [comment]
Another alternative is to use the [Flatpak version](https://flathub.org/apps/io.github.nuttyartist.notes), which still uses Qt 6.4.

But yeah, this is a pretty annoying bug. :(

**Comment by guihkx** [comment]
I think I found a consistent way to trigger/prevent the freezing.

**Steps to force the app to freeze:**

1. Open [`Settings.ini`](https://github.com/nuttyartist/notes#database-path) with a text editor
2. Replace the entire `windowGeometry` line by this:

    ```ini
    windowGeometry=@ByteArray(\x1\xd9\xd0\xcb\0\x3\0\0\0\0\x1\x96\0\0\0\xa3\0\0\x5\xe7\0\0\x3u\0\0\x1\x96\0\0\0\xc0\0\0\x5\xe7\0\0\x3u\0\0\0\0\0\0\0\0\a\x80\0\0\x1\x96\0\0\0\xc0\0\0\x5\xe7\0\0\x3u)
    ```
3. Maximizing/restoring the window should freeze the app.

**Steps to prevent the app from freezing:**

1. Open [`Settings.ini`](https://github.com/nuttyartist/notes#database-path) with a text editor 
2. Replace the entire `windowGeometry` line by this:

    ```ini
    windowGeometry=@ByteArray(\x1\xd9\xd0\xcb\0\x3\0\0\0\0\x1#\0\0\0\xe5\0\0\x6\"\0\0\x3\x96\0\0\x1#\0\0\x1\x2\0\0\x6\"\0\0\x3\x96\0\0\0\0\0\0\0\0\a\x80\0\0\x1#\0\0\x1\x2\0\0\x6\"\0\0\x3\x96)
    ```
3. Maximizing/restoring the window should *not* freeze the app.

I haven't looked into what these values mean yet, but these steps have worked on my Linux machine. In any case, it'd be good to get another confirmation on this... Additionally, and unfortunately, I still haven't found a way to reproduce this with a minimal example so we can report it to Qt devs. =(

If you're going to test this, just remember to do it with a Qt 6.5.x build. This is not reproducible with any of the official binaries we release, since they still use Qt 6.4.x.

**Comment by nuttyartist** [comment]
I haven't tested this yet, have you tried printing the window geometry to see if there's anything unusual with the problematic one?

**Comment by guihkx** [comment]
No yet, no. I'll definitely look into that though - unless someone beats me to it.

**Comment by guihkx** [comment]
I forgot to post an update on this, but in short:

I [reported](https://bugreports.qt.io/browse/QTBUG-115459) this upstream, and I also managed to create a minimal, reproducible example that triggers the bug only with Qt 6.5+, but unfortunately there hasn't been an official response yet.

Instead of waiting on upstream, perhaps we can find a workaround in our code base instead...

If someone wants to give this a look, essentially, the app freezes in [`CustomDocument::setDocumentPadding()`](https://github.com/nuttyartist/notes/blob/438b7f5b2d7fb3d7306c664c75fa364790124606/src/customdocument.cpp#L26), after calling [`QAbstractScrollArea::setViewportMargins()`](https://doc.qt.io/qt-6/qabstractscrollarea.html#setViewportMargins).

The call stack looks basically like this:

1. We listen for the window resize event [here](https://github.com/nuttyartist/notes/blob/438b7f5b2d7fb3d7306c664c75fa364790124606/src/mainwindow.cpp#L3957)
2. We call `MainWindow::setCurrentFontBasedOnTypeface()` [here](https://github.com/nuttyartist/notes/blob/438b7f5b2d7fb3d7306c664c75fa364790124606/src/mainwindow.cpp#L3972)
3. We call `MainWindow::alignTextEditText()` [here](https://github.com/nuttyartist/notes/blob/438b7f5b2d7fb3d7306c664c75fa364790124606/src/mainwindow.cpp#L1143)
4. We call `MainWindow::setupTextEditStyleSheet()` [here](https://github.com/nuttyartist/notes/blob/438b7f5b2d7fb3d7306c664c75fa364790124606/src/mainwindow.cpp#L1201)
5. We call `CustomDocument::setDocumentPadding()` [here](https://github.com/nuttyartist/notes/blob/438b7f5b2d7fb3d7306c664c75fa364790124606/src/mainwindow.cpp#L1186)
6. We call `setViewportMargins()` [here](https://github.com/nuttyartist/notes/blob/438b7f5b2d7fb3d7306c664c75fa364790124606/src/customdocument.cpp#L28C5-L28C23) <-- Freezing occurs

**Comment by nuttyartist** [comment]
Weird... I can't reproduce this on Windows 10 with Qt 6.5.2 MinGW 64bit.

**Comment by guihkx** [comment]
@nuttyartist Did you try reproducing it with default settings? And with actual text inside the `textEdit` widget?

Also, can you try reproducing it with [this small project](https://bugreports.qt.io/secure/attachment/150496/repro.zip)?

Just build it as usual (`cmake -B build`), run the program, and simply maximize and unmaximize the window.

It should freeze it after that...

**Comment by nuttyartist** [comment]
> Did you try reproducing it with default settings? And with actual text inside the textEdit widget?

Yes, straight from master. And yes, with some text.

> Also, can you try reproducing it with [this small project](https://bugreports.qt.io/secure/attachment/150496/repro.zip)?

Yes, this one hangs.

Weird.

**Comment by nuttyartist** [comment]
https://github.com/nuttyartist/notes/assets/16375940/de7ca59c-9b8c-434c-8ca6-47b6fb868023

**Comment by guihkx** [comment]
Good news: This bug has been fixed in Qt 6.6.3!

I'll leave this open until 6.6.3 is released, which should be at the end of February, and then we can upgrade our official packages to use that version.

**Comment by guihkx** [comment]
Qt 6.6.3 has been out for quite a while now, and since the v2.2.1 binaries we distribute still use Qt 6.4 (where this issue is not present), I think we can close this.
